File: sdk/reproit-windows/src/ReproIt.Core/Engine.cs
// Platform-agnostic core of the SDK. Holds the state-graph state machine, label
// reduction, batching and payload building. Has NO WPF / WinUI dependency so the
// behavior is unit-testable on any host; the Windows layer (ReproItClient) feeds
// it raw node data and supplies the wall clock + transport.
//
// Mirrors sdk/reproit-android/.../Engine.kt and sdk/reproit-web.js exactly: the
// edge/error state machine, the PII-safe context map (SetContext / SetContexts /
// Identify with the SHA-256 "uid" hash) and the {appId, sentAt, ctx?, events}
// batch envelope are byte-for-byte the same wire shape as the other SDKs and the
// cloud's ingest endpoint (POST /v1/events).

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Security.Cryptography;
using System.Text;

namespace ReproIt.Core
{
    /// <summary>The result of evaluating an app invariant: it either HELD or was
    /// VIOLATED with a message. Use <see cref="Ok"/> / <see cref="Fail"/> to build
    /// one. A predicate that throws is also a violation (message = the exception
    /// text), mirroring the web SDK's "throws => violated" contract.</summary>
    public sealed class InvariantResult
    {
        public bool Held { get; }
        public string Message { get; }

        private InvariantResult(bool held, string message)
        {
            Held = held;
            Message = message;
        }

        public static InvariantResult Ok()
        {
            return new InvariantResult(true, null);
        }

        public static InvariantResult Fail(string message = "")
        {
            return new InvariantResult(false, message ?? string.Empty);
        }
    }

    /// <summary>One observed accessibility node, as the platform layer reads it.</summary>
    public sealed class RawNode
    {
        /// <summary>AutomationProperties.Name ?? text, already non-null (empty if neither).</summary>
        public string Name { get; }

        /// <summary>True if the underlying element is invokable/clickable.</summary>
        public bool Tappable { get; }

        public RawNode(string name, bool tappable)
        {
            Name = name ?? string.Empty;
            Tappable = tappable;
        }
    }

    /// <summary>Reduced snapshot: a signature and capped unique labels.</summary>
    public sealed class Snapshot
    {
        public string Sig { get; }
        public List<string> Labels { get; }

        public Snapshot(string sig, List<string> labels)
        {
            Sig = sig;
            Labels = labels;
        }
    }

    /// <summary>One step in the graph trail kept for repros.</summary>
    public sealed class Step
    {
        public string Sig { get; }
        public string Action { get; }
        public string Label { get; }

        public Step(string sig, string action, string label = null)
        {
            Sig = sig;
            Action = action;
            Label = label;
        }

        public IDictionary<string, object> ToMap()
        {
            var m = new Dictionary<string, object>();
            m["sig"] = Sig;
            m["action"] = Action;
            if (!string.IsNullOrEmpty(Label))
            {
                m["label"] = Label;
            }
            return m;
        }
    }

    public sealed class Engine
    {
        private readonly ReproItConfig _cfg;
        private readonly Func<long> _now;
        private readonly Func<string, bool> _transport;
        private readonly Action<string> _log;

        private readonly List<IDictionary<string, object>> _queue = new List<IDictionary<string, object>>();
        private readonly List<Step> _path = new List<Step>();
        private readonly object _queueLock = new object();
        private readonly object _ctxLock = new object();
        private string _currentSig;
        private string _pendingAction;
        private string _pendingLabel;

        // PII-safe context dimensions sent with each batch (the "which users" answer).
        // Insertion-ordered and merged in place; included as the batch envelope's
        // "ctx" field (only when non-empty), exactly like the other SDKs.
        private readonly Dictionary<string, object> _context = new Dictionary<string, object>();

        // App-declared invariants, keyed by id (registration idempotent). Inert in
        // production; evaluated only under the fuzzer (see ReportInvariants).
        private readonly Dictionary<string, Func<InvariantResult>> _invariants =
            new Dictionary<string, Func<InvariantResult>>(StringComparer.Ordinal);

        public Engine(
            ReproItConfig cfg,
            Func<long> now = null,
            Func<string, bool> transport = null,
            Action<string> log = null)
        {
            _cfg = cfg;
            _now = now ?? (() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
            _transport = transport ?? (_ => true);
            _log = log;
        }

        public int QueueSize
        {
            get { lock (_queueLock) { return _queue.Count; } }
        }

        public string CurrentSignature()
        {
            return _currentSig;
        }

        /// <summary>Read-only snapshot of the current context dimensions (for tests / debug).</summary>
        public IDictionary<string, object> Context()
        {
            lock (_ctxLock) { return new Dictionary<string, object>(_context); }
        }

        /// <summary>Set a single PII-safe context dimension (e.g. role, plan, a count bucket).</summary>
        public void SetContext(string key, object value)
        {
            lock (_ctxLock) { _context[key] = value; }
        }

        /// <summary>Merge several context dimensions at once.</summary>
        public void SetContexts(IDictionary<string, object> values)
        {
            lock (_ctxLock)
            {
                foreach (var kv in values)
                {
                    _context[kv.Key] = kv.Value;
                }
            }
        }

        /// <summary>Attach a hashed user id (so the cloud can group "these N users hit
        /// it" without storing identity) plus optional context dimensions. The raw
        /// userId is never stored or sent; only a SHA-256 hex prefix is kept as "uid".</summary>
        public void Identify(string userId, IDictionary<string, object> context = null)
        {
            lock (_ctxLock)
            {
                _context["uid"] = HashUid(userId);
                if (context != null)
                {
                    foreach (var kv in context)
                    {
                        _context[kv.Key] = kv.Value;
                    }
                }
            }
        }

        private static string HashUid(string userId)
        {
            using (var sha = SHA256.Create())
            {
                byte[] digest = sha.ComputeHash(Encoding.UTF8.GetBytes(userId));
                var sb = new StringBuilder(16);
                // First 8 bytes -> 16 lowercase hex chars (matches the other SDKs' 16-char prefix).
                for (int i = 0; i < 8; i++)
                {
                    sb.Append(digest[i].ToString("x2", CultureInfo.InvariantCulture));
                }
                return sb.ToString();
            }
        }

        /// <summary>Accessible name reduction shared with the snapshot path. Trim, take
        /// the first line, drop empties and labels longer than MaxLabelLen.</summary>
        public string CleanLabel(string raw)
        {
            if (raw == null)
            {
                return null;
            }
            string trimmed = raw.Trim();
            int nl = trimmed.IndexOf('\n');
            string first = (nl >= 0 ? trimmed.Substring(0, nl) : trimmed).Trim();
            if (first.Length == 0 || first.Length > _cfg.MaxLabelLen)
            {
                return null;
            }
            return first;
        }

        /// <summary>Reduce a flat list of visible nodes (pre-order) plus the captured
        /// structural tree into a Snapshot. The signature is STRUCTURAL: the canonical
        /// descriptor of the tree prefixed by the screen anchor (route), byte-identical
        /// to the Rust oracle and the other SDKs (docs/signature.md). Localized text
        /// never enters the hash. The flat node list is used only for the display-only
        /// "labels" field; labels do NOT affect the signature.</summary>
        public Snapshot Reduce(IList<RawNode> nodes, Node tree, string anchor = null)
        {
            var seen = new List<string>();
            var seenSet = new HashSet<string>(StringComparer.Ordinal);
            foreach (var n in nodes)
            {
                string label = CleanLabel(n.Name);
                if (label == null)
                {
                    continue;
                }
                if (seenSet.Add(label))
                {
                    seen.Add(label);
                }
            }
            string sig = Signature.Of(anchor, tree);
            var capped = seen.Count > _cfg.MaxLabels ? seen.GetRange(0, _cfg.MaxLabels) : seen;
            return new Snapshot(sig, capped);
        }

        /// <summary>Record the action a tap implies; consumed by the next state change.</summary>
        public void NoteTap(string selector, string label)
        {
            _pendingAction = !string.IsNullOrEmpty(selector) ? "tap:" + selector : "tap:?";
            _pendingLabel = label;
        }

        /// <summary>Record an explicit navigation action.</summary>
        public void NoteNav()
        {
            _pendingAction = "nav";
            _pendingLabel = null;
        }

        /// <summary>Observe a reduced snapshot. If the signature changed (or this is the
        /// first observation), record an edge and advance the current state.</summary>
        /// <summary>Register an app invariant: a predicate the app declares that must
        /// hold in EVERY visited state. It returns <see cref="InvariantResult.Ok"/>
        /// when it holds, or <see cref="InvariantResult.Fail"/> (or throws) when it is
        /// violated. Under the fuzzer the SDK evaluates every registered invariant on
        /// each Observe and reports the failures for the runner to turn into
        /// `invariant` findings; in production the registry is INERT (evaluated only
        /// under the fuzzer), so it is zero-overhead until a run reproduces it.
        /// Registration is idempotent by id, so re-registering an id replaces it.</summary>
        public Engine Invariant(string id, Func<InvariantResult> predicate)
        {
            if (!string.IsNullOrEmpty(id) && predicate != null)
            {
                _invariants[id] = predicate;
            }
            return this;
        }

        public void Observe(Snapshot snap, string firstAction = "load")
        {
            // App-invariant oracle (SDK-self-triggered): under the fuzzer, evaluate
            // the app's registered predicates against this state and report failures
            // on stderr, which the UIA runner scrapes. No-op in production.
            ReportInvariants(snap.Sig);
            string cur = _currentSig;
            if (cur == null)
            {
                _currentSig = snap.Sig;
                EmitEdge(null, firstAction, snap, true);
                return;
            }
            if (snap.Sig == cur)
            {
                return;
            }
            string action = _pendingAction ?? "auto";
            string label = !_cfg.RedactLabels ? _pendingLabel : null;
            _pendingAction = null;
            _pendingLabel = null;
            EmitEdge(cur, action, snap, true, label);
            _currentSig = snap.Sig;
        }

        private void EmitEdge(string from, string action, Snapshot to, bool append, string label = null)
        {
            if (append)
            {
                _path.Add(new Step(from ?? string.Empty, action, label));
                if (_path.Count > _cfg.PathCap)
                {
                    _path.RemoveAt(0);
                }
            }
            var ev = new Dictionary<string, object>();
            ev["kind"] = "edge";
            if (from != null)
            {
                ev["from"] = from;
            }
            ev["action"] = action;
            if (!string.IsNullOrEmpty(label))
            {
                ev["label"] = label;
            }
            ev["to"] = to.Sig;
            if (!_cfg.RedactLabels)
            {
                ev["labels"] = to.Labels;
            }
            ev["t"] = _now();
            Enqueue(ev);
        }

        /// <summary>Evaluate one predicate. Returns null when it HELD, or a failure
        /// message string when it was VIOLATED (the returned message, or the thrown
        /// exception's text).</summary>
        private static string EvalInvariant(Func<InvariantResult> predicate)
        {
            try
            {
                var r = predicate();
                if (r == null || r.Held)
                {
                    return null;
                }
                return r.Message ?? string.Empty;
            }
            catch (Exception ex)
            {
                return string.IsNullOrEmpty(ex.Message) ? ex.GetType().Name : ex.Message;
            }
        }

        /// <summary>Evaluate every registered invariant and, ONLY under the fuzzer (the
        /// REPROIT_UNDER_FUZZER env var the UIA runner sets on the launched child, the
        /// fuzzer-detection gate), write one marker line
        ///   REPROIT_INVARIANT {"sig":"&lt;sig&gt;","items":[{"id","message"}...]}
        /// listing the VIOLATED invariants to stderr, which the runner scrapes and
        /// re-emits as EXPLORE:INVARIANT. Silent when the registry is empty or every
        /// invariant held (no empty-items line); inert in production (env var unset).</summary>
        private void ReportInvariants(string sig)
        {
            if (_invariants.Count == 0
                || string.IsNullOrEmpty(Environment.GetEnvironmentVariable("REPROIT_UNDER_FUZZER")))
            {
                return;
            }
            var items = new List<object>();
            foreach (var kv in _invariants)
            {
                string message = EvalInvariant(kv.Value);
                if (message != null)
                {
                    var it = new Dictionary<string, object>();
                    it["id"] = kv.Key;
                    it["message"] = message;
                    items.Add(it);
                }
            }
            if (items.Count == 0)
            {
                return;
            }
            var payload = new Dictionary<string, object>();
            payload["sig"] = sig ?? string.Empty;
            payload["items"] = items;
            Console.Error.WriteLine("REPROIT_INVARIANT " + Json.Encode(payload));
            Console.Error.Flush();
        }

        /// <summary>Record an error event carrying the current signature and graph path.
        /// stack is capped to 8 lines. Returns the event (useful for tests / for the
        /// caller to flush synchronously before a crash).</summary>
        public IDictionary<string, object> RecordError(
            string message,
            IList<string> stack,
            string source = "",
            int line = 0,
            IDictionary<string, object> context = null)
        {
            var ev = new Dictionary<string, object>();
            ev["kind"] = "error";
            // A genuine uncaught error IS the `crash` oracle firing; tag it so
            // the cloud can gate ingest on oracle-grade findings.
            ev["oracle"] = "crash";
            ev["sig"] = _currentSig ?? string.Empty;
            var pathOut = new List<object>();
            foreach (var s in _path)
            {
                pathOut.Add(s.ToMap());
            }
            // Include the in-flight action: a click whose handler throws
            // synchronously sets _pendingAction but crashes before its debounced
            // observe records it, so the bare path stops one step short of the bug.
            if (!string.IsNullOrEmpty(_pendingAction))
            {
                pathOut.Add(new Dictionary<string, object>
                {
                    ["sig"] = _currentSig ?? string.Empty,
                    ["action"] = _pendingAction,
                    ["label"] = !_cfg.RedactLabels ? _pendingLabel : null,
                });
            }
            ev["path"] = pathOut;
            ev["message"] = message;
            var cappedStack = new List<object>();
            int take = Math.Min(8, stack.Count);
            for (int i = 0; i < take; i++)
            {
                cappedStack.Add(stack[i]);
            }
            ev["stack"] = cappedStack;
            ev["source"] = source;
            ev["line"] = line;
            if (context != null && context.Count > 0)
            {
                ev["context"] = context;
            }
            ev["t"] = _now();
            Enqueue(ev);
            return ev;
        }

        private void Enqueue(IDictionary<string, object> ev)
        {
            try
            {
                _cfg.OnEvent?.Invoke(ev);
            }
            catch
            {
                // a faulty dev hook must never break telemetry.
            }
            if (_cfg.Endpoint == null)
            {
                if (_cfg.OnEvent == null)
                {
                    _log?.Invoke("reproit " + Json.Encode(ev));
                }
                return;
            }
            lock (_queueLock)
            {
                _queue.Add(ev);
            }
            if (QueueSize >= 50)
            {
                Flush();
            }
        }

        /// <summary>Build the batch body for the given queued events (without draining).</summary>
        public string BuildBatch(IList<IDictionary<string, object>> events)
        {
            var envelope = new Dictionary<string, object>();
            envelope["appId"] = _cfg.AppId;
            envelope["sentAt"] = _now();
            // Include "ctx" only when non-empty, placed BEFORE "events" to match the
            // other SDKs' envelope order.
            Dictionary<string, object> ctx;
            lock (_ctxLock)
            {
                ctx = _context.Count == 0 ? null : new Dictionary<string, object>(_context);
            }
            if (ctx != null)
            {
                envelope["ctx"] = ctx;
            }
            var evList = new List<object>();
            foreach (var e in events)
            {
                evList.Add(e);
            }
            envelope["events"] = evList;
            return Json.Encode(envelope);
        }

        /// <summary>Drain the queue and ship it via the transport. On failure the batch
        /// is re-queued ahead of newer events for one retry (mirrors the other SDKs).</summary>
        public void Flush()
        {
            if (_cfg.Endpoint == null)
            {
                lock (_queueLock) { _queue.Clear(); }
                return;
            }
            List<IDictionary<string, object>> batch;
            lock (_queueLock)
            {
                if (_queue.Count == 0)
                {
                    return;
                }
                batch = new List<IDictionary<string, object>>(_queue);
                _queue.Clear();
            }
            string body = BuildBatch(batch);
            bool ok;
            try
            {
                ok = _transport(body);
            }
            catch
            {
                ok = false;
            }
            if (!ok)
            {
                lock (_queueLock)
                {
                    _queue.InsertRange(0, batch);
                }
            }
        }
    }
}

File: sdk/reproit-android/src/main/kotlin/com/reproit/android/Engine.kt
package com.reproit.android

/**
 * Platform-agnostic core of the SDK. Holds the state-graph state machine, label reduction, batching
 * and payload building. Has NO `android.*` imports so the behavior is unit-testable on the host
 * JVM; the Android layer ([ReproIt]) feeds it raw node data and supplies the wall clock +
 * transport.
 *
 * Mirrors `sdk/reproit-web.js` and `sdk/reproit_flutter/lib/reproit_flutter.dart`.
 */

/** One observed accessibility node, as the platform layer reads it. */
data class RawNode(
  /** contentDescription ?? text, already non-null (empty string if neither). */
  val name: String,
  /** true if the underlying view is clickable (`View.isClickable`). */
  val tappable: Boolean,
)

/** Reduced snapshot: a signature and capped unique display labels. */
data class Snapshot(val sig: String, val labels: List<String>)

/** One step in the graph trail kept for repros. */
data class Step(val sig: String, val action: String, val label: String? = null) {
  fun toMap(): Map<String, Any?> = mapOf("sig" to sig, "action" to action, "label" to label)
}

data class PendingStep(val action: String, val label: String? = null) {
  fun toStep(sig: String): Step = Step(sig, action, label)
}

class Engine(
  private val cfg: ReproItConfig,
  /** Wall clock in epoch milliseconds; injectable for tests. */
  private val now: () -> Long = { System.currentTimeMillis() },
  /**
   * Transport for a serialized batch body. Receives the JSON string; returns true on success. The
   * default is a no-op (the Android layer wires HTTP).
   */
  private val transport: (String) -> Boolean = { true },
  /** Optional plain-text logger used only when there is no endpoint/onEvent. */
  private val log: ((String) -> Unit)? = null,
) {
  private val queue = ArrayList<Map<String, Any?>>()
  private val path = ArrayList<Step>()
  private var currentSig: String? = null
  private var pending: PendingStep? = null
  private var batchSequence = 0L

  /**
   * App-declared invariants (see [ReproIt.invariant]): predicates that must hold in every visited
   * state. A plain SDK-owned store, idempotent by id and insertion-ordered; INERT in production
   * (never evaluated) and consulted only when the Android layer detects it is running under the
   * fuzzer, so registration is zero-overhead. Kept in the pure-Kotlin core so the registry +
   * evaluation are host-testable; the fuzzer gate + log emission live in the android layer
   * ([ReproIt]).
   */
  private val invariants = LinkedHashMap<String, () -> Boolean>()

  /**
   * PII-safe context dimensions sent with each batch (the "which users" answer). Insertion-ordered
   * and merged in place; included in each finding event's `context` field, exactly like the
   * Flutter/web SDKs.
   */
  private val context = LinkedHashMap<String, Any?>()

  init {
    val build = LinkedHashMap<String, String>()
    cfg.buildVersion?.takeIf { it.isNotBlank() }?.let { build["version"] = it }
    cfg.buildCommit?.takeIf { it.isNotBlank() }?.let { build["commit"] = it }
    if (build.isNotEmpty()) context["build"] = build
  }

  val queueSize: Int
    get() = synchronized(queue) { queue.size }

  fun currentSignature(): String? = currentSig

  /** Read-only snapshot of the current context dimensions (for tests / debug). */
  fun context(): Map<String, Any?> = synchronized(context) { LinkedHashMap(context) }

  /** Set a single PII-safe context dimension (e.g. role, plan, a count bucket). */
  fun setContext(key: String, value: Any?) {
    synchronized(context) { context[key] = value }
  }

  /** Merge several context dimensions at once. */
  fun setContexts(values: Map<String, Any?>) {
    synchronized(context) { context.putAll(values) }
  }

  /**
   * Attach a hashed user id (so the cloud can group "these N users hit it" without storing
   * identity) plus optional context dimensions. The raw [userId] is never stored or sent; only a
   * SHA-256 hex prefix is kept as `uid`. Mirrors the Flutter SDK's `identify`.
   */
  fun identify(userId: String, context: Map<String, Any?>? = null) {
    synchronized(this.context) {
      this.context["uid"] = hashUid(userId)
      if (context != null) this.context.putAll(context)
    }
  }

  private fun hashUid(userId: String): String {
    val digest =
      java.security.MessageDigest.getInstance("SHA-256").digest(userId.toByteArray(Charsets.UTF_8))
    val sb = StringBuilder(32)
    // First 8 bytes -> 16 lowercase hex chars (matches Flutter's substring(0, 16)).
    for (i in 0 until 8) sb.append("%02x".format(digest[i]))
    return sb.toString()
  }

  /**
   * Accessible name reduction shared with the snapshot path. Trim, take the first line, drop
   * empties and labels longer than [ReproItConfig.maxLabelLen].
   */
  fun cleanLabel(raw: String): String? {
    val first = raw.trim().substringBefore('\n').trim()
    if (first.isEmpty() || first.length > cfg.maxLabelLen) return null
    return first
  }

  /**
   * Reduce a flat list of visible nodes (pre-order: parents before children) plus the captured
   * structural [tree] into a [Snapshot].
   *
   * The signature is STRUCTURAL: the canonical descriptor of [tree] prefixed by the screen [anchor]
   * (route), byte-identical to the Rust oracle and the other SDKs (docs/signature.md). Localized
   * text never enters the hash.
   *
   * The flat [nodes] list is used only for the display-only `labels` field (`map --show`): it
   * dedupes labels and caps to [ReproItConfig.maxLabels]. Labels do NOT affect the signature.
   */
  fun reduce(nodes: List<RawNode>, tree: Signature.Node, anchor: String? = null): Snapshot {
    val seen = LinkedHashSet<String>()
    for (n in nodes) {
      val label = cleanLabel(n.name)
      if (label == null) continue
      seen.add(label)
    }
    val unique = seen.toList()
    val sig = Signature.of(anchor, tree)
    return Snapshot(sig, unique.take(cfg.maxLabels))
  }

  /**
   * Register an app invariant, idempotent by [id] (re-registering an id replaces it). [test]
   * returns true when the invariant HOLDS; returning false or throwing marks it VIOLATED (a thrown
   * exception's message becomes the finding message). Mirrors the web SDK's `ReproIt.invariant`.
   * Evaluation is gated by the fuzzer in the android layer, so this is inert in production.
   */
  fun registerInvariant(id: String, test: () -> Boolean) {
    synchronized(invariants) { invariants[id] = test }
  }

  /**
   * Evaluate every registered invariant; return one `{id,message}` entry per VIOLATED invariant
   * (held ones are omitted). Each predicate is isolated so one throwing predicate cannot suppress
   * the others. Does NOT apply the fuzzer gate (that lives in [ReproIt]); host-testable.
   */
  fun evaluateInvariants(): List<Map<String, Any?>> {
    val snapshot = synchronized(invariants) { LinkedHashMap(invariants) }
    val out = ArrayList<Map<String, Any?>>()
    for ((id, test) in snapshot) {
      var ok = true
      var message = ""
      try {
        ok = test()
      } catch (t: Throwable) {
        ok = false
        message = t.message ?: t.toString()
      }
      if (!ok) out.add(linkedMapOf("id" to id, "message" to message))
    }
    return out
  }

  /**
   * The `REPROIT_INVARIANT` marker line to log when one or more invariants are violated, else null
   * (silent). The emitted sig is left empty ("") so the mobile runner substitutes the sig it is
   * currently on. The android layer logs this to logcat only under the fuzzer; the Rust core is
   * never touched.
   */
  fun invariantMarker(): String? {
    val items = evaluateInvariants()
    if (items.isEmpty()) return null
    val obj = LinkedHashMap<String, Any?>()
    obj["sig"] = ""
    obj["items"] = items
    return "REPROIT_INVARIANT " + Json.encode(obj)
  }

  /** Record the action a tap implies; consumed by the next state change. */
  fun noteTap(selector: String?, label: String?) {
    pending =
      PendingStep(
        action = if (selector != null && selector.isNotEmpty()) "tap:$selector" else "tap:?",
        label = label,
      )
  }

  /** Record an explicit navigation action. */
  fun noteNav() {
    pending = PendingStep("nav")
  }

  /**
   * Observe a reduced snapshot. If the signature changed (or this is the first observation), record
   * an edge and advance the current state. `firstAction` is the action used for the very first
   * observed state ("load").
   */
  fun observe(snap: Snapshot, firstAction: String = "load") {
    val cur = currentSig
    if (cur == null) {
      currentSig = snap.sig
      emitEdge(from = null, action = firstAction, to = snap, append = true)
      return
    }
    if (snap.sig == cur) return
    val step = pending ?: PendingStep("auto")
    pending = null
    emitEdge(from = cur, step = step, to = snap, append = true)
    currentSig = snap.sig
  }

  private fun emitEdge(from: String?, action: String, to: Snapshot, append: Boolean) {
    emitEdge(from, PendingStep(action), to, append)
  }

  private fun emitEdge(from: String?, step: PendingStep, to: Snapshot, append: Boolean) {
    if (append) {
      path.add(step.toStep(from ?: ""))
      if (path.size > cfg.pathCap) path.removeAt(0)
    }
    val ev = LinkedHashMap<String, Any?>()
    ev["kind"] = "edge"
    if (from != null) ev["from"] = from
    ev["action"] = step.action
    if (!cfg.redactLabels && step.label != null) ev["label"] = step.label
    ev["to"] = to.sig
    if (!cfg.redactLabels) ev["labels"] = to.labels
    ev["t"] = now()
    enqueue(ev)
  }

  /**
   * Record an error event carrying the current signature and graph path. [stack] is capped to 8
   * lines. Returns the event (useful for tests / for the caller to flush synchronously before a
   * crash).
   */
  fun recordError(
    message: String,
    stack: List<String>,
    source: String = "",
    line: Int = 0,
    /**
     * PII-safe tier-3 on-error context (input fingerprints under `context.fingerprint`). Omitted
     * from the wire when null/empty.
     */
    context: Map<String, Any?>? = null,
  ): Map<String, Any?> {
    val ev = LinkedHashMap<String, Any?>()
    ev["kind"] = "error"
    // A genuine uncaught error IS the `crash` oracle firing; tag it so the
    // cloud can gate ingest on oracle-grade findings.
    ev["oracle"] = "crash"
    ev["sig"] = currentSig ?: ""
    // Include the in-flight action: a click whose handler throws synchronously
    // sets `pendingAction` but crashes before its debounced observe records it,
    // so the bare path stops one step short of the crashing tap.
    val pathOut = path.map { it.toMap() }.toMutableList()
    pending?.let { pathOut.add(it.toStep(currentSig ?: "").toMap()) }
    ev["path"] = pathOut
    ev["message"] = message
    ev["stack"] = stack.take(8)
    ev["source"] = source
    ev["line"] = line
    if (context != null && context.isNotEmpty()) ev["context"] = context
    ev["t"] = now()
    enqueue(ev)
    return ev
  }

  /** Capture the current structural state as a tester-observed bug. */
  fun captureBug(context: Map<String, Any?>? = null): Map<String, Any?>? {
    val sig = currentSig ?: return null
    val pathOut = path.map { it.toMap() }.toMutableList()
    pending?.let { pathOut.add(it.toStep(sig).toMap()) }
    val trigger = (pathOut.lastOrNull()?.get("action") as? String) ?: "load"
    val ev =
      linkedMapOf<String, Any?>(
        "kind" to "error",
        "oracle" to "tester-capture",
        "sig" to sig,
        "path" to pathOut,
        "message" to "Tester observed a bug in this state",
        "findingIdentity" to
          linkedMapOf(
            "oracle" to "tester-capture",
            "invariant" to "tester-observed-failure",
            "kind" to "structural-state",
            "message" to "",
            "frame" to "",
            "trigger" to trigger,
            "boundary" to sig,
          ),
        "t" to now(),
      )
    if (context != null && context.isNotEmpty()) ev["context"] = context
    enqueue(ev)
    return ev
  }

  fun captureContractBug(identity: String, message: String): Map<String, Any?>? {
    val sig = currentSig ?: return null
    val pathOut = path.map { it.toMap() }.toMutableList()
    pending?.let { pathOut.add(it.toStep(sig).toMap()) }
    val trigger = (pathOut.lastOrNull()?.get("action") as? String) ?: "load"
    val ev =
      linkedMapOf<String, Any?>(
        "kind" to "error",
        "oracle" to "invariant",
        "sig" to sig,
        "path" to pathOut,
        "message" to message,
        "findingIdentity" to
          linkedMapOf(
            "oracle" to "invariant",
            "invariant" to identity,
            "kind" to "structural-contract",
            "message" to message,
            "frame" to "",
            "trigger" to trigger,
            "boundary" to sig,
          ),
        "t" to now(),
      )
    enqueue(ev)
    return ev
  }

  private fun enqueue(ev: Map<String, Any?>) {
    try {
      cfg.onEvent?.invoke(ev)
    } catch (_: Throwable) {}
    if (cfg.endpoint == null) {
      if (cfg.onEvent == null) log?.invoke("reproit " + Json.encode(ev))
      return
    }
    synchronized(queue) {
      queue.add(ev)
      if (queue.size >= 50) {
        // flush inline to bound memory; the timer also flushes.
      }
    }
    if (queueSize >= 50) flush()
  }

  /** Build a strict version 1 event batch for the currently queued capture records. */
  fun buildBatch(events: List<Map<String, Any?>>): String {
    val sentAt = now()
    val sequence = synchronized(this) { ++batchSequence }
    val batchId = "sdk-$sentAt-$sequence"
    val ctx = synchronized(context) { if (context.isEmpty()) null else LinkedHashMap(context) }
    val frames =
      events.mapIndexed { index, event ->
        linkedMapOf<String, Any?>(
          "runId" to batchId,
          "sequence" to index + 1,
          "scope" to linkedMapOf("domain" to "shared"),
          "event" to protocolEvent(event, ctx ?: emptyMap()),
        )
      }
    return Json.encode(
      linkedMapOf<String, Any?>(
        "version" to 1,
        "batchId" to batchId,
        "appId" to cfg.appId,
        "frames" to frames,
        "evidence" to emptyList<Any?>(),
      )
    )
  }

  private fun protocolEvent(
    event: Map<String, Any?>,
    batchContext: Map<String, Any?>,
  ): Map<String, Any?> {
    if (event["kind"] == "edge") {
      return linkedMapOf(
        "kind" to "graph-edge",
        "from" to (event["from"] ?: "∅"),
        "action" to (event["action"] ?: "auto"),
        "to" to (event["to"] ?: "?"),
      )
    }
    if (event["kind"] != "error") {
      return linkedMapOf("kind" to "stream-defect", "reason" to "invalid-event")
    }

    val path = (event["path"] as? List<*>)?.filterIsInstance<Map<String, Any?>>() ?: emptyList()
    val message = event["message"]?.toString() ?: ""
    val identity =
      event["findingIdentity"]
        ?: linkedMapOf(
          "oracle" to (event["oracle"]?.toString() ?: "crash"),
          "invariant" to "no-exception",
          "kind" to "exception",
          "message" to structuralMessage(message),
          "frame" to "",
          "trigger" to (path.lastOrNull()?.get("action")?.toString() ?: ""),
          "boundary" to null,
        )
    val findingContext = LinkedHashMap(batchContext)
    @Suppress("UNCHECKED_CAST")
    (event["context"] as? Map<String, Any?>)?.let { findingContext.putAll(it) }
    return linkedMapOf(
      "kind" to "finding",
      "signature" to (event["sig"]?.toString() ?: "?"),
      "message" to message,
      "identity" to identity,
      "path" to
        path.map { step ->
          linkedMapOf(
            "signature" to (step["sig"] ?: "?"),
            "action" to (step["action"] ?: "auto"),
            "label" to step["label"],
          )
        },
      "context" to findingContext,
    )
  }

  private fun structuralMessage(message: String): String =
    message.replace(Regex("([\"']).*?\\1"), "<q>").replace(Regex("[0-9][0-9.,]*"), "#")

  /**
   * Drain the queue and ship it via [transport]. On failure the batch is re-queued ahead of newer
   * events for one retry (mirrors the Flutter SDK).
   */
  fun flush() {
    if (cfg.endpoint == null) {
      synchronized(queue) { queue.clear() }
      return
    }
    val batch: List<Map<String, Any?>>
    synchronized(queue) {
      if (queue.isEmpty()) return
      batch = ArrayList(queue)
      queue.clear()
    }
    val body = buildBatch(batch)
    val ok =
      try {
        transport(body)
      } catch (_: Throwable) {
        false
      }
    if (!ok) {
      synchronized(queue) { queue.addAll(0, batch) }
    }
  }
}
